File: 2-D Dynamic Programming/808. Soup Servings/solution.py
class Solution:
    def soupServings(self, n: int) -> float:
        if n > 5000: return 1
        @lru_cache(None)
        def dfs(a, b):
            if a > 0 and b <= 0: return 0
            if a <= 0 and b > 0: return 1
            if a <= 0 and b <= 0: return 0.5

            prob = 0
            prob += dfs(a-100, b) * 0.25
            prob += dfs(a-75, b-25) * 0.25
            prob += dfs(a-50, b-50) * 0.25
            prob += dfs(a-25, b-75) * 0.25
            return prob
        return dfs(n, n)

File: 2-D Dynamic Programming/808. Soup Servings/solution.cpp
class Solution {
private:
    double servings[4][2] = {{4, 0}, {3, 1}, {2, 2}, {1, 3}};
public:
    double soupServings(int n) {
        int m = ceil(n / 25.0);

        unordered_map<int, unordered_map<int, double>> dp;

        function<double(int, int)> dfs = [&](int A, int B) -> double {
            if (A <= 0 && B <= 0) {
                return 0.5;
            }
            if (A <= 0) {
                return 1;
            }
            if (B <= 0) {
                return 0;
            }
            if (dp[A].count(B)) {
                return dp[A][B];
            }

            for (const auto& [dA, dB] : servings) {
                dp[A][B] += 0.25 * dfs(A-dA, B-dB);
            }
            return dp[A][B];
        };

        for (int k = 1; k <= m; k++) {
            if (dfs(k, k) > 1 - 1e-5) { // 0.99999... almost 1.0
                return 1;
            }
        }
        return dfs(m, m);
    }
};